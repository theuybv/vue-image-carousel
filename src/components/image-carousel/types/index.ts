import { InjectionKey } from "vue";

export enum AspectRatio {
  "16/9" = 16 / 9,
  "9/16" = 9 / 16,
  "3/2" = 3 / 2,
  "3/4" = 3 / 4,
  "4/3" = 4 / 3,
  "1/1" = 1,
}

export type CarouselImage = {
  imageSrc: string;
  thumbSrc: string;
  alt?: string;
};

export type ThumbsInOutView = {
  index: number;
  element: HTMLElement;
  isInView: boolean;
};

export type ImageCarouselProviderProps = {
  readonly images: CarouselImage[];
  readonly thumbsCount: number;
  readonly thumbsGap: number;
  readonly thumbsWidth: number;
  readonly currentImage: CarouselImage;
  readonly imageAspectRatio: AspectRatio;
  readonly thumbAspectRatio: AspectRatio;
  readonly imageMaxHeight: number;
  readonly thumbsScrollDelay: number;
  readonly hasIndicator: boolean;
  readonly autoPlayMode: "none" | "loop" | "default";
  imageContainerWidth: number;
  thumbElements: HTMLElement[];
  thumbsContainerElement: HTMLElement;
  currentIndex: number;
  allLoadedThumbs: Promise<CarouselImage[]>;
};

export const ImageCarouselKey =
  Symbol() as InjectionKey<ImageCarouselProviderProps>;
