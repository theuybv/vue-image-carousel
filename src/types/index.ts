import { InjectionKey } from "vue";

export enum AspectRatio {
  "16/9" = 16 / 9,
  "9/16" = 9 / 16,
  "3/2" = 3 / 2,
  "3/4" = 3 / 4,
  "4/3" = 4 / 3,
  "1/1" = 1,
}

export type CarouselImage = {
  imageSrc: string;
  thumbSrc: string;
  alt?: string;
};

export type ThumbsInOutView = {
  index: number;
  element: HTMLElement;
  isInView: boolean;
};

export type ImageCarouselProvider = {
  images: CarouselImage[];
  imageContainerWidth: number;
  updateImageContainerWidth: (value: number) => void;
  updateCurrentIndex: (value: number) => void;
  updateThumbsContainerElement: (value: HTMLElement) => void;
  updateThumbElements: (value: HTMLElement[]) => void;
  thumbElements: HTMLElement[];
  thumbsContainerElement: HTMLElement;
  thumbsCount: number;
  thumbsGap: number;
  thumbsWidth: number;
  currentImage: CarouselImage;
  currentIndex: number;
  imageAspectRatio: AspectRatio;
  thumbAspectRatio: AspectRatio;
  imageMaxHeight: number;
  allLoadedThumbs: Promise<CarouselImage[]>;
};

export const ImageCarouselKey = Symbol() as InjectionKey<ImageCarouselProvider>;
