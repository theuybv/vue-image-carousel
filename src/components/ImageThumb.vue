<script setup lang="ts">

import {AspectRatio, CarouselImage} from "../types";

export type ImageThumbProps = {
  image: CarouselImage,
  aspectRatio: AspectRatio
  width: number
  onImageThumbLoaded: () => void
}
const {image, aspectRatio, width, onImageThumbLoaded} = defineProps<ImageThumbProps>()

</script>

<template>
  <button class="ImageThumb">
    <img @load="onImageThumbLoaded" :src="image.imageSrc" :alt="image.alt"/>
  </button>
</template>

<style scoped>
.ImageThumb {
  border: none;
  cursor: pointer;
  display: flex;
  width: v-bind(width+ 'px');
  padding: 0;
  margin: 0;
  aspect-ratio: v-bind(aspectRatio);
}

img {
  width: 100%;
  height: 100%;
  aspect-ratio: v-bind(aspectRatio);
  background: black;
  object-fit: cover;
}
</style>
