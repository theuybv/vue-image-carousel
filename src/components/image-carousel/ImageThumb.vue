<script setup lang="ts">

import {AspectRatio, CarouselImage} from "../../types";

export type ImageThumbProps = {
  image: CarouselImage,
  aspectRatio: AspectRatio
  width: number
}
const {image, aspectRatio, width} = defineProps<ImageThumbProps>()

</script>

<template>
  <button class="ImageThumb" :style="{width: width + 'px', aspectRatio: aspectRatio.toString()}">
    <img :src="image.imageSrc" :alt="image.alt" :style="{aspectRatio: aspectRatio.toString()}"/>
  </button>
</template>

<style scoped>
.ImageThumb {
  border: none;
  cursor: pointer;
  display: flex;
  padding: 0;
  margin: 0;
}

img {
  width: 100%;
  height: 100%;
  background: black;
  object-fit: cover;
}
</style>
