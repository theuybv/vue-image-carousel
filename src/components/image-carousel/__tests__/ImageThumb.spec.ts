import ImageThumb, { ImageThumbProps } from "../ImageThumb.vue";
import { render } from "@testing-library/vue";
import { AspectRatio } from "../types";

test("render component with correct html elements", () => {
  const props: ImageThumbProps = {
    active: true,
    aspectRatio: AspectRatio["1/1"],
    image: {
      thumbSrc: "",
      imageSrc: "",
      alt: "",
    },
    width: 100,
  };
  const { getByRole } = render(ImageThumb, { props });

  const button = getByRole("button");
  const img = getByRole("img");
  expect(button).toBeInTheDocument();
  expect(img).toBeInTheDocument();
  expect(button).toMatchSnapshot();
});
