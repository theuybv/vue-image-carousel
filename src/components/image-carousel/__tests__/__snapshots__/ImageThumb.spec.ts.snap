// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component with correct html elements 1`] = `
<button
  class="box-content cursor-pointer flex flex-col p-0 m-0"
  style="width: 100px;"
>
  <img
    alt=""
    class="hover:scale-[0.98] transition-all w-full h-full bg-black object-cover"
    src=""
  />
  <div
    class="animate-fade h-[4px] w-full bg-red-600"
  />
</button>
`;
